discovery/gossiper.go

```diff
@@ -527,19 +529,42 @@ func (d *deDupedAnnouncements) addMsg(message networkMsg) {
 			msg.Flags,
 		}
 
+		oldTimestamp := uint32(0)
 		mws, ok := d.channelUpdates[deDupKey]
-		if !ok {
+		if ok {
+			// If we already have seen this message, record its
+			// timestamp.
+			oldTimestamp = mws.msg.(*lnwire.ChannelUpdate).Timestamp
+		}
+
+		// If we already had this message with a strictly newer
+		// timestamp, then we'll just discard the message we got.
+		if oldTimestamp > msg.Timestamp {
+			return
+		}
+
+		// If the message we just got is newer than what we previously
+		// have seen, or this is the first time we see it, then we'll
+		// add it to our map of announcements.
+		if oldTimestamp < msg.Timestamp {
 			mws = msgWithSenders{
 				msg:     msg,
 				senders: make(map[routing.Vertex]struct{}),
 			}
+
+			// We'll mark the sender of the message in the
+			// senders map.
 			mws.senders[sender] = struct{}{}
 
 			d.channelUpdates[deDupKey] = mws
 
 			return
 		}
 
+		// Lastly, if we had seen this exact message from before, with
+		// the same timestamp, we'll add the sender to the map of
+		// senders, such that we can skip sending this message back in
+		// the next batch.
```

Comment:
Should we also validate that the messages themselves are equivalent? What happens if peer maliciously distributes two different channel updates with the same timestamp?

Reply:
Not sure if it matters? We'll just forward one of the every trickle anyway. Basic validation is done in `processNetworkAnnouncement` when they come in.

Reply:
Good to know, thanks :)